test/cpp/interop/backend_metrics_lb_policy.cc

```diff
@@ -0,0 +1,240 @@
+//
+//
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include <grpc/support/port_platform.h>
+
+#include "test/cpp/interop/backend_metrics_lb_policy.h"
+
+#include "google/protobuf/util/message_differencer.h"
+
+#include "src/core/lib/iomgr/pollset_set.h"
+
+namespace grpc {
+namespace testing {
+
+namespace {
+
+using grpc_core::CoreConfiguration;
+using grpc_core::LoadBalancingPolicy;
+using grpc_core::MakeRefCounted;
+using grpc_core::OrphanablePtr;
+using grpc_core::RefCountedPtr;
+
+constexpr absl::string_view kBackendMetricsLbPolicyName =
+    "test_backend_metrics_load_balancer";
+
+const std::string kMetricsTrackerAttribute = "orca_metrics_tracker";
+
+absl::optional<xds::data::orca::v3::OrcaLoadReport>
+BackendMetricDataToOrcaLoadReport(
+    const grpc_core::BackendMetricData* backend_metric_data) {
+  if (backend_metric_data == nullptr) {
+    return absl::nullopt;
+  }
+  xds::data::orca::v3::OrcaLoadReport load_report;
+  load_report.set_cpu_utilization(backend_metric_data->cpu_utilization);
+  load_report.set_mem_utilization(backend_metric_data->mem_utilization);
+  load_report.set_rps_fractional(backend_metric_data->qps);
+  for (const auto& p : backend_metric_data->request_cost) {
+    std::string name(p.first);
+    (*load_report.mutable_request_cost())[name] = p.second;
+  }
+  for (const auto& p : backend_metric_data->utilization) {
+    std::string name(p.first);
+    (*load_report.mutable_utilization())[name] = p.second;
+  }
+  return load_report;
+}
+
+class BackendMetricsLbPolicy : public LoadBalancingPolicy {
+ public:
+  explicit BackendMetricsLbPolicy(Args args)
+      : LoadBalancingPolicy(std::move(args), /*initial_refcount=*/2) {
+    Args delegate_args;
+    delegate_args.work_serializer = work_serializer();
+    delegate_args.args = channel_args();
+    delegate_args.channel_control_helper =
+        std::make_unique<Helper>(RefCountedPtr<BackendMetricsLbPolicy>(this));
+    load_report_tracker_ = delegate_args.args.GetPointer<LoadReportTracker>(
```

Comment:
Suggest moving this to the top of the function, before you start to build delegate_args, since it's really not related to constructing the delegate policy.  Note that it's fine to call `channel_args()` twice, since that method returns a const reference.

Also, please add an assertion that `load_report_tracker_` is non-null.

Reply:
Done